File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/MapLibreException.java
package org.maplibre.nativeffi;

import java.util.Objects;

/** Base unchecked exception for errors reported by the native MapLibre C ABI. */
public class MapLibreException extends RuntimeException {
  private final MapLibreStatus status;
  private final int nativeStatusCode;
  private final String diagnostic;

  public MapLibreException(MapLibreStatus status, int nativeStatusCode, String diagnostic) {
    super(message(status, nativeStatusCode, diagnostic));
    this.status = Objects.requireNonNull(status, "status");
    this.nativeStatusCode = nativeStatusCode;
    this.diagnostic = diagnostic == null ? "" : diagnostic;
  }

  public MapLibreStatus status() {
    return status;
  }

  public int nativeStatusCode() {
    return nativeStatusCode;
  }

  public String diagnostic() {
    return diagnostic;
  }

  public static MapLibreException forStatus(
      MapLibreStatus status, int nativeStatusCode, String diagnostic) {
    return switch (status) {
      case INVALID_ARGUMENT -> new InvalidArgumentException(nativeStatusCode, diagnostic);
      case INVALID_STATE -> new InvalidStateException(nativeStatusCode, diagnostic);
      case WRONG_THREAD -> new WrongThreadException(nativeStatusCode, diagnostic);
      case UNSUPPORTED -> new UnsupportedFeatureException(nativeStatusCode, diagnostic);
      case NATIVE_ERROR -> new NativeErrorException(nativeStatusCode, diagnostic);
      case OK, UNKNOWN -> new MapLibreException(status, nativeStatusCode, diagnostic);
    };
  }

  private static String message(MapLibreStatus status, int nativeStatusCode, String diagnostic) {
    var detail =
        diagnostic == null || diagnostic.isBlank() ? "No native diagnostic available." : diagnostic;
    return "%s (%d): %s".formatted(status, nativeStatusCode, detail);
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/error/MaplibreException.kt
package org.maplibre.nativeffi.error

/** Base unchecked exception for errors reported by the native MapLibre C ABI. */
public open class MaplibreException
internal constructor(
  public val status: MaplibreStatus,
  public val nativeStatusCode: Int,
  public val diagnostic: String = "",
) : RuntimeException(message(status, nativeStatusCode, diagnostic)) {
  internal companion object {
    /** Creates the stable exception subtype for a native status category. */
    fun forStatus(
      status: MaplibreStatus,
      nativeStatusCode: Int,
      diagnostic: String = "",
    ): MaplibreException =
      when (status) {
        MaplibreStatus.INVALID_ARGUMENT -> InvalidArgumentException(nativeStatusCode, diagnostic)
        MaplibreStatus.INVALID_STATE -> InvalidStateException(nativeStatusCode, diagnostic)
        MaplibreStatus.WRONG_THREAD -> WrongThreadException(nativeStatusCode, diagnostic)
        MaplibreStatus.UNSUPPORTED -> UnsupportedFeatureException(nativeStatusCode, diagnostic)
        MaplibreStatus.NATIVE_ERROR -> NativeErrorException(nativeStatusCode, diagnostic)
        else -> MaplibreException(status, nativeStatusCode, diagnostic)
      }
  }
}

private fun message(status: MaplibreStatus, nativeStatusCode: Int, diagnostic: String): String {
  val detail = diagnostic.ifBlank { "No native diagnostic available." }
  return "${statusLabel(status)} ($nativeStatusCode): $detail"
}

private fun statusLabel(status: MaplibreStatus): String =
  when (status) {
    MaplibreStatus.OK -> "OK"
    MaplibreStatus.INVALID_ARGUMENT -> "INVALID_ARGUMENT"
    MaplibreStatus.INVALID_STATE -> "INVALID_STATE"
    MaplibreStatus.WRONG_THREAD -> "WRONG_THREAD"
    MaplibreStatus.UNSUPPORTED -> "UNSUPPORTED"
    MaplibreStatus.NATIVE_ERROR -> "NATIVE_ERROR"
    else -> "MaplibreStatus(${status.nativeCode})"
  }
